openfasoc/generators/glayout/glayout/flow/blocks/elementary/cascode_common_source/cascode_common_source.py

```diff
@@ -0,0 +1,407 @@
+
+	Returns:
+		Component: a cascode common source amplifier component object
+	"""
+	top_level = Component("cascode common source amplifier")
+	# Create the transistors
+	print(f"Creating ", device, " devices with these parameters: ",m1_fingers,m1_multipliers, m2_fingers, m2_multipliers)
+	if device in ['nmos', 'nfet']:
+		fet_M1=nmos(pdk,
+					fingers=m1_fingers,
+					multipliers = m1_multipliers,
+					with_tie=True,
+					with_dummy=False, #with_dummy,
+					with_substrate_tap=False,
+					**kwargs)
+		fet_M2=nmos(pdk,
+					fingers=m2_fingers,
+					multipliers = m2_multipliers,
+					with_tie=True,
+					with_dummy=False, #with_dummy,
+					with_substrate_tap=False,
+					**kwargs)
+	elif device in ['pmos', 'pfet']:
+		fet_M1=pmos(pdk,
+					fingers=m1_fingers,
+					multipliers = m1_multipliers,
+					with_tie=True,
+					with_dummy=False, #with_dummy,
+					with_substrate_tap=False,
+					**kwargs)
+		fet_M2=pmos(pdk,
+					fingers=m2_fingers,
+					multipliers = m2_multipliers,
+					with_tie=True,
+					with_dummy=False, #with_dummy,
+					with_substrate_tap=False,
+					**kwargs)
+	print("FETS are instantiated now")
+	
+	# Added references to the two FETs within the component level
+	M1_ref = top_level << fet_M1
+	M2_ref = top_level << fet_M2
+
+	top_level.add(M1_ref)
+	top_level.add(M2_ref)
+	# Placement
+	M1_ref_centre_coord = prec_ref_center(fet_M1)
+	M2_ref_centre_coord = prec_ref_center(fet_M2)
+
+	# Place the devices Horizontally ('H') or Vertically('V') based on placement selection
+	place_devices='V' 
+	if place_devices in ['lateral', 'horizontal', 'H']:
+		M2_ref.movex(0.75*(evaluate_bbox(M1_ref)[0]+evaluate_bbox(M2_ref)[0]))
+	if place_devices in ['vertical', 'V']:
+		M2_ref.movey(0.5*(evaluate_bbox(M1_ref)[1]+evaluate_bbox(M2_ref)[1]))
+		
+	maxmet_sep = pdk.util_max_metal_seperation()
+
+	# Routing and Port definitions
+	# if place_devices in ['lateral', 'horizontal', 'H']:
+	# 	top_level << straight_route(pdk, M1_ref.ports["multiplier_0_drain_W"], M2_ref.ports["multiplier_0_source_E"])
+	# if  place_devices in ['vertical', 'V']:
+	# 	top_level << c_route(pdk, M1_ref.ports["multiplier_0_drain_E"], M2_ref.ports["multiplier_0_source_E"])
+		
+
+	# ************* Adding the suffix after the routing generates the prefix after routing.
+	#### top_level.unlock()
+	top_level.add_ports(M1_ref.get_ports_list(), prefix="M1_")
+	top_level.add_ports(M2_ref.get_ports_list(), prefix="M2_")
+
+	# Shorting the Source of M2 to the Drain of M1.
+	int_net_short = top_level << c_route(pdk, 
+											top_level.ports["M2_source_E"], 
+											top_level.ports["M1_drain_E"], 
+											extension=4*maxmet_sep, 
+											viaoffset=False)
+	top_level.add_ports(int_net_short.get_ports_list(), prefix="INT")
+
+	# add well
+	if device in ['nmos', 'nfet']:
+		print(f"NMOS device")
+		# add a pwell 
+		top_level.add_padding(layers = (pdk.get_glayer("pwell"),), default = pdk.get_grule("pwell", "active_tap")["min_enclosure"], )
```

Comment:
Both of these statements only change the value `nwell` or `pwell`, code duplication can be reduced.